set helper (#594)
emberjs/rfcs: text/0000-set-helper.md

- Start Date: 2020-02-20
- Relevant Team(s): Ember.js
- RFC PR: (after opening the RFC PR, update this with a link to it and update the file name)
- Tracking: (leave this empty)

# `set` helper

## Summary

This RFC introduces a new `set` helper like the one provided by
[`ember-simple-set-helper`](https://github.com/pzuraq/ember-simple-set-helper).

> The `set` helper will always return a function that will change a value.

Reviewer: ```suggestion
The `set` helper will always return a function that will change value.
```

```hbs
<SelectCountry @update={{set this.country}} />
```

The new `set` helper can be used instead of the old `{{fn (mut ..)}}` combination.

> However this RFC does not deprecate `{{mut}}`.

Reviewer 2: Why *not* deprecate `{{mut}}`, preferably with a timeline where there are one or more minor releases between the time when `{{set}}` is added and `{{mut}}` is deprecated so that addons can migrate without causing unactionable noise for consumers.

If there's a reason we *cannot* or *should not* deprecate `{{mut}}`, I think (a) we should spell that out here, but also (b) it undercuts the motivation for this RFC. We end up in a spot where there are two ways of doing something, one of them likely preferred/recommended but the other still supported and without a clear indication that you should move away from it to the new preferred/recommended approach. If there are gaps in functionality from `{{mut}}` which we believe need to be covered, it's fine to deliver them separately, but that should be expressly called out in the text of this RFC.

Author: Honestly I think there are *maybe* gaps, be we need to find them. We have no replacement for the *mutable attribute* syntax. I think long term we should deprecate `mut`, but this will be a first step in that direction. A future RFC should deprecate mut, and maybe additional RFCs are required for the gaps. I will try to make this clear in the RFC. Thanks for pointing it out.

## Motivation

The existing `mut` helper is quite helpful, but behaves often in an unexpected way.

When used to pass an argument as in `<MyComponent @foo={{mut this.bar}}>` or
`{{my-component foo=(mut this.bar)}}` it will do nothing for a glimmer component
but provide a box with a `value` and an `update` function on the private `attrs`
property of a classic component.

However the more common use-case is to pass the result of `mut` to either the 
`{{action}}` or the `{{fn}}` helper:

```hbs
<SelectCountry @update={{fn (mut this.country)}} />
<button {{on "click" (fn (mut this.name) "")}}>delete name</button>
```

However many people would expect `{{fn (mut this.foo)}}` to be the same as

> `{{mut this.foo}}` in all cases, which can lead to confusion.

Reviewer 2: Could you expand this with more detail on what exactly the difference is? This is a distinction I've loaded up into my head a number of times, but since I make a point *not* to reach for `{{mut}}` I inevitably forget it—and it's not clear to me what the actual difference is *in practice and effect* between these two. If you could show in these examples how they differ in behavior, that would help not only me but everyone else in evaluating this RFC!

Author: I tried to explain the problem a bit. I would love some feedback if its understandable.

So this proposal proposes a new `set` helper that will work as `mut` wrapped

> inside `fn`.

Reviewer 2: Nit: phrasing—this should probably say something like:

> This proposal proposes a new `set` helper with the semantics users *expect* from `mut`: the way it behaves when wrapped with the `fn` helper.

## Detailed design

The new `set` helper can be called with one or two arguments.

### one argument version

The one argument version returns a function that receives one argument `value` and
will set the property passed as argument to `value`.
```hbs
{{set context.property}}
```

```js
return function(arg) {
  context.property = arg;
};
```

The first argument must *always* have a context. So `{{set foo}}` or `{{set @foo}}` is not valid.
This aligns with [RFC 0308](https://github.com/emberjs/rfcs/blob/master/text/0308-deprecate-property-lookup-fallback.md).

The fact that the `set` helper somehow needs to know the *context* and the *name* of the
argument it receives makes it magic syntax that can not be used by a userland helper without
a AST transformation. But the need for the helper seems big enough to justify this, especially

> when the same magic syntax is already used by the `mut` helper.

Reviewer 2: What about this scenario?

```hbs
{{#each @values as |value|}}
  <FancyInput @onChange={{set value}} />
{{/each}}
```

This seems like a *very* common use case, and one that the current proposed design would rule out.

Reviewer 2: Perhaps this is part of the reason this doesn't fully replace `{{mut}}`! If so, that's probably worth making explicit in the text of the RFC.

Reviewer 2: I just realized that this exact behavior probably doesn't work *at all* regardless of this. A better example would be something like:

```hbs
{{#each @users as |user|}}
  <FancyInput @onChange={{set user.name}} />
{{/each}}
```

The alternative, less magical version as suggested in another comment would just be:

```hbs
{{#each @users as |user|}}
  <FancyInput @onChange={{set user 'name'}} />
{{/each}}
```

I could see an argument either way here in terms of the brevity-vs.-magic tradeoff. Apologies for the confusion!

Author: exactly. `{{set user.name}}` would work. Also this does *not* work currently (just tested it):

```
{{#each @values as |value|}}
  <FancyInput @onChange={{fn (mut value)}} />
{{/each}}
```

I generally would want `{{set user.name}}` and not `{{set user "name"}}`. But indeed, both would work.

> when the same magic syntax is already used by the `mut` helper.

Reviewer 2: I actually disagree with the idea that because `mut` does this magic, so should `set`. Indeed, part of the reason that `mut` is confusing is that it *does* do magical things and doesn't behave like other helpers. Accordingly, I suggest—

- that you add an alternative where we *only* add the two-argument version
- that you mark this design question as an Unanswered Question

Reviewer 3: Strongly agree. Especially I think that `set` helper should have a similar syntax as the [existing `get` helper](https://api.emberjs.com/ember/3.16/classes/Ember.Templates.helpers/methods/get?anchor=get), which is `{{get context "prop"}}`.

> ### two argument version

Reviewer: ```suggestion
### two arguments version
```

> The two argument version `{{set context.property value}}` is a shorthand for

Reviewer: ```suggestion
The two arguments version `{{set context.property value}}` is a shorthand for
```

Reviewer 2: The most correct English grammar here is actually hyphenation:

```suggestion
The two-argument version `{{set context.property value}}` is a shorthand for
```

`{{fn (set context.property) value}}` and will always perform the same.

### dynamic attribute binding

The existing `mut` helper allows to receive the result of the `get` helper to create a 
[dynamic attribute binding](https://guides.emberjs.com/release/components/built-in-components/#toc_binding-dynamic-attribute):

```hbs
<Input @value={{mut (get this.person this.field)}}
```

This use-case is *not* covered by the new `set` helper, however the existing `mut` helper

> can still be used for this.

Reviewer 2: This might be an argument for the context-and-key alternative, `{{set context fieldName}}`, as that would handily solve this and generalize to the other cases as well.

## How we teach this

> We can teach teach it the same way as we used to tech `mut`.

Reviewer: ```suggestion
We can teac it the same way as we used to teach `mut`.
```

> We must replace all occurances of `mut` in the guides and replace it by `set`.

Reviewer: ```suggestion
We must replace all occurrences of `mut` in the guides and replace them by `set`.
```

## Drawbacks

> It increases API surface, and provides two helpers for the same functionallity.

Reviewer: ```suggestion
It increases the API surface and provides two helpers for the same functionality.
```

However this can be mitigated by clearly communicating that `set` should be
used in favor of `mut`.

## Alternatives

We could leave everything as it is and encourage people to use the `{{fn (mut` combination.
Or we could try to change `mut` to behave like the `set` proposed here with an *optional feature*.

## Unresolved questions

> Do we want the two argument version? We could just go with the one argument version.

Reviewer: ```suggestion
Do we want the two arguments version? We could just go with the one argument version.
```
